pkg/standalone-utils/contracts/ProtocolFeeSweeper.sol

```diff
@@ -0,0 +1,200 @@
+ * convert, and forward them to the final recipient.
+ */
+contract ProtocolFeeSweeper is IProtocolFeeSweeper, SingletonAuthentication {
+    using SafeERC20 for IERC20;
+
+    /// @notice All pool tokens are ERC20, so this contract should not handle ETH.
+    error CannotReceiveEth();
+
+    /// @notice The fee recipient is invalid.
+    error InvalidFeeRecipient();
+
+    // Preferred token for receiving protocol fees. Passed to the fee burner as the target of fee token swaps.
+    IERC20 private _targetToken;
+
+    // Final destination of the collected protocol fees.
+    address private _feeRecipient;
+
+    // Address of the current "burner" contract (i.e., swapping fee tokens to the target).
+    IProtocolFeeBurner private _protocolFeeBurner;
+
+    // The default configuration on deployment simply forwards all fee tokens to the `feeRecipient`.
+    constructor(IVault vault, address feeRecipient) SingletonAuthentication(vault) {
+        _setFeeRecipient(feeRecipient);
+    }
+
+    /// @inheritdoc IProtocolFeeSweeper
+    function sweepProtocolFees(address pool) external {
+        IERC20[] memory poolTokens = getVault().getPoolTokens(pool);
+        uint256 numTokens = poolTokens.length;
+
+        // There could be tokens "left over" from uncompleted burns from previous sweeps. Only process the "new"
+        // tokens from the current withdrawal.
+        uint256[] memory existingBalances = new uint256[](numTokens);
+        for (uint256 i = 0; i < numTokens; ++i) {
+            existingBalances[i] = poolTokens[i].balanceOf(address(this));
+        }
+
+        // Withdraw protocol fees to this contract. Note that governance will need to grant this contract permission
+        // to call `withdrawProtocolFees` on the `ProtocolFeeController.
+        getProtocolFeeController().withdrawProtocolFees(pool, address(this));
+
+        // Convert the given tokens to the target token (if enabled), and forward to the fee recipient.
+        IProtocolFeeBurner feeBurner = _protocolFeeBurner;
+        IERC20 targetToken = _targetToken;
+        address recipient = _feeRecipient;
+
+        // There must be a burner contract and valid target token to enable converting fees to the preferred currency.
+        // If these were not set, fall back on simply transferring all fee tokens directly to the recipient.
+        bool canBurn = targetToken != IERC20(address(0)) && feeBurner != IProtocolFeeBurner(address(0));
+
+        for (uint256 i = 0; i < numTokens; ++i) {
+            IERC20 feeToken = poolTokens[i];
+            uint256 withdrawnTokenBalance = feeToken.balanceOf(address(this)) - existingBalances[i];
+
+            // If no balance, nothing to do.
+            if (withdrawnTokenBalance == 0) {
+                continue;
+            }
+
+            // If this is already the target token (or we haven't enabled burning), just forward directly.
+            if (canBurn && feeToken != targetToken) {
```

Comment:
I think it should be split into two functions. One that calls burn, and one that does safeTransfer. It will be more convenient when we integrate CowSwapBurner. WDYT?

Reply:
The "burn" one would still have to fall back to forwarding if the contract/target were undefined, though. If we want to collect and forward even though we've assigned a burner, I suppose we could - but if this is being called by governance, at that point they could just call withdraw directly and not have to transfer tokens twice.

The safeTransfer mode is just a fallback in case the burner's undefined. In future PRs we will likely have a choice of burners, or a mapping of pools to burners, or something like that, to make it more flexible. At that point, "no entry" in the map for a pool would mean don't use a burner and just forward it.

We're only constrained by the CowSwapBurner if we're directly reusing it. If we're rewriting it, we can make things simpler.